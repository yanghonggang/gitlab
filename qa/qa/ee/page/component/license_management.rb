# frozen_string_literal: true

module QA
  module EE
    module Page
      module Component
        module LicenseManagement
          extend QA::Page::PageConcern

          def self.prepended(base)
            super

            base.class_eval do
              view 'app/assets/javascripts/reports/components/report_item.vue' do
                element :report_item_row
              end

              view 'app/assets/javascripts/reports/components/issue_status_icon.vue' do
                element :icon_status, ':data-qa-selector="`status_${status}_icon`" ' # rubocop:disable QA/ElementWithPattern
              end

              view 'ee/app/assets/javascripts/vue_shared/license_compliance/mr_widget_license_report.vue' do
                element :license_report_widget
                element :manage_licenses_button
              end
            end
          end

          def has_approved_license?(name)
            within_element(:report_item_row, text: name) do
              has_element?(:status_success_icon, wait: 1)
            end
          end

          def has_denied_license?(name)
            within_element(:report_item_row, text: name) do
              has_element?(:status_failed_icon, wait: 1)
            end
          end

          def click_license(name)
            within_element(:license_report_widget) do
              click_on name
            end
            wait_for_animated_element(:license_management_modal)
          end

          def click_manage_licenses_button
            previous_page = page.current_url
            within_element(:license_report_widget) do
              click_element :manage_licenses_button
            end
            wait_until(max_duration: 15, reload: false) do
              page.current_url != previous_page
            end
          end
        end
      end
    end
  end
end
