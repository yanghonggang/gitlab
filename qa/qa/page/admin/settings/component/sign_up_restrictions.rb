# frozen_string_literal: true

module QA
  module Page
    module Admin
      module Settings
        module Component
          class SignUpRestrictions < Page::Base
            view 'app/views/admin/application_settings/_signup.html.haml' do
              element :require_admin_approval_after_user_signup_checkbox
              element :save_changes_button
            end

            def require_admin_approval_after_user_signup
              check_element :require_admin_approval_after_user_signup_checkbox
              click_element :save_changes_button
            end
          end
        end
      end
    end
  end
end
