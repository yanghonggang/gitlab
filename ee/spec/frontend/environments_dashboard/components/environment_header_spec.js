import { GlLink, GlBadge, GlIcon } from '@gitlab/ui';
import { shallowMount, mount } from '@vue/test-utils';
import component from 'ee/environments_dashboard/components/dashboard/environment_header.vue';
import ReviewAppLink from '~/vue_merge_request_widget/components/review_app_link.vue';

describe('Environment Header', () => {
  let wrapper;
  let propsData;

  beforeEach(() => {
    propsData = {
      environment: {
        environment_path: '/enivronment/1',
        name: 'staging',
        external_url: 'http://example.com',
      },
    };
  });

  afterEach(() => {
    wrapper.destroy();
  });

  describe('renders name and link to app', () => {
    beforeEach(() => {
      wrapper = mount(component, {
        propsData,
      });
    });

    it('renders the environment name', () => {
      expect(wrapper.find('.js-environment-name').text()).toBe(propsData.environment.name);
    });

    it('renders a link to the environment page', () => {
      expect(wrapper.find(GlLink).attributes('href')).toBe(propsData.environment.environment_path);
    });

    it('does not show a badge with the number of environments in the folder', () => {
      expect(wrapper.find(GlBadge).exists()).toBe(false);
    });

    it('renders a link to the external app', () => {
      expect(wrapper.find(ReviewAppLink).attributes('href')).toBe(
        propsData.environment.external_url,
      );
    });

    it('matches the snapshot', () => {
      expect(wrapper.element).toMatchSnapshot();
    });
  });

  describe('with environments grouped into a folder', () => {
    beforeEach(() => {
      propsData.environment.size = 5;
      propsData.environment.within_folder = true;
      propsData.environment.name = 'review/testing';

      wrapper = shallowMount(component, {
        propsData,
      });
    });

    it('shows a badge with the number of other environments in the folder', () => {
      const expected = propsData.environment.size.toString();
      expect(wrapper.find(GlBadge).text()).toBe(expected);
    });

    it('shows an icon stating the environment is one of many in a folder', () => {
      expect(wrapper.find(GlIcon).attributes('name')).toBe('information');
      expect(wrapper.find(GlIcon).attributes('title')).toMatch(/last updated environment/);
    });

    it('matches the snapshot', () => {
      expect(wrapper.element).toMatchSnapshot();
    });
  });

  describe('has errors', () => {
    beforeEach(() => {
      propsData.hasErrors = true;

      wrapper = shallowMount(component, {
        propsData,
      });
    });

    it('matches the snapshot', () => {
      expect(wrapper.element).toMatchSnapshot();
    });
  });

  describe('has a failed pipeline', () => {
    beforeEach(() => {
      propsData.hasPipelineFailed = true;

      wrapper = shallowMount(component, {
        propsData,
      });
    });

    it('matches the snapshot', () => {
      expect(wrapper.element).toMatchSnapshot();
    });
  });
});
