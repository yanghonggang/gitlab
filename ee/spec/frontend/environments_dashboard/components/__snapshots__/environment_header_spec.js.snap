// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Environment Header has a failed pipeline matches the snapshot 1`] = `
<div
  class="card-header border-0 py-2 d-flex align-items-center dashboard-card-header-failed"
>
  <div
    class="flex-grow-1 block-truncated"
  >
    <gl-link-stub
      class="js-environment-link cgray"
      href="/enivronment/1"
      title="staging"
    >
      <span
        class="js-environment-name bold"
      >
         staging
      </span>
    </gl-link-stub>
     
    <!---->
  </div>
   
  <review-app-link-stub
    cssclass="btn btn-default btn-sm"
    display="[object Object]"
    link="http://example.com"
    size="medium"
  />
</div>
`;

exports[`Environment Header has errors matches the snapshot 1`] = `
<div
  class="card-header border-0 py-2 d-flex align-items-center dashboard-card-header-warning"
>
  <div
    class="flex-grow-1 block-truncated"
  >
    <gl-link-stub
      class="js-environment-link cgray"
      href="/enivronment/1"
      title="staging"
    >
      <span
        class="js-environment-name bold"
      >
         staging
      </span>
    </gl-link-stub>
     
    <!---->
  </div>
   
  <review-app-link-stub
    cssclass="btn btn-default btn-sm"
    display="[object Object]"
    link="http://example.com"
    size="medium"
  />
</div>
`;

exports[`Environment Header renders name and link to app matches the snapshot 1`] = `
<div
  class="card-header border-0 py-2 d-flex align-items-center bg-light"
>
  <div
    class="flex-grow-1 block-truncated"
  >
    <a
      class="gl-link js-environment-link cgray"
      href="/enivronment/1"
      title="staging"
    >
      <span
        class="js-environment-name bold"
      >
         staging
      </span>
    </a>
     
    <!---->
  </div>
   
  <a
    class="btn btn-default btn-md gl-button btn btn-default btn-sm"
    data-track-event="open_review_app"
    data-track-label="review_app"
    href="http://example.com"
    rel="noopener noreferrer nofollow"
    target="_blank"
    title=""
  >
    <!---->
     
    <!---->
      
    <span
      class="gl-button-text"
    >
      
  View app 
      <svg
        class="fgray gl-icon s16"
        data-testid="external-link-icon"
      >
        <use
          href="#external-link"
        />
      </svg>
    </span>
  </a>
</div>
`;

exports[`Environment Header with environments grouped into a folder matches the snapshot 1`] = `
<div
  class="card-header border-0 py-2 d-flex align-items-center bg-light"
>
  <div
    class="flex-grow-1 block-truncated"
  >
    <gl-link-stub
      class="js-environment-link cgray"
      href="/enivronment/1"
      title="review/testing"
    >
      <span
        class="js-environment-name bold"
      >
         review/testing
      </span>
    </gl-link-stub>
     
    <gl-badge-stub
      class="dashboard-card-icon"
      size="md"
      variant="muted"
    >
      5
    </gl-badge-stub>
  </div>
   
  <gl-icon-stub
    class="dashboard-card-icon"
    name="information"
    size="16"
    title="You are looking at the last updated environment"
  />
</div>
`;
